### PR TITLE
Allows ethereals to ghost out of crystallization

### PR DESCRIPTION
**Pull request by wesoda25**
## About The Pull Request

Ethereals can now ghost during crystallization, instead of being stuck in their body for the duration. Just press the ghost button! It'll break the crystal and also make you unable to re-enter your body (like normal ghosting).

## Why It's Good For The Game

QOL. Very annoying to have to sit out the duration if you don't want to be brought back into the round.

I considered allowing ethereals to opt out entirely via a preference, but I think crystallization was an intended quirk for ethereals that everyone has to live with. Thoughts influenced by what floyd said in this thread: https://tgstation13.org/phpBB/viewtopic.php?f=9&t=29859

## Changelog

:cl:
qol: ethereals can now press the ghost button to cancel crystallization and permanently vacate their body
/:cl:

### CODE REVIEW

**Comment by tralezab** [comment]
i approve of this change

File: code/modules/mob/dead/observer/observer.dm [hunk 1]
```diff
@@ -283,8 +283,14 @@ Works together with spawning an observer, noted above.
 /mob/proc/ghostize(can_reenter_corpse = TRUE)
 	if(key)
 		if(key[1] != "@") // Skip aghosts.
-			if(HAS_TRAIT(src, TRAIT_CORPSELOCKED) && can_reenter_corpse) //If you can re-enter the corpse you can't leave when corpselocked
-				return
+			if(HAS_TRAIT(src, TRAIT_CORPSELOCKED))
+				if(can_reenter_corpse) //If you can re-enter the corpse you can't leave when corpselocked
+					return
+				else //otherwise, continues the ghosting process
```
Comment (tralezab) [inline code/modules/mob/dead/observer/observer.dm:289]: there is no reason to have the below block of code under and else case, considering the if check ends the proc with a return

File: code/modules/mob/dead/observer/observer.dm [hunk 2]
```diff
@@ -312,8 +318,15 @@ This is the proc mobs get to turn into a ghost. Forked from ghostize due to comp
...
-		return TRUE
+		if(isethereal(usr))
+			var/response = tgui_alert(usr, "Are you sure you want to ghost? If you ghost whilst crystallizing you cannot re-enter your body!", "Confirm Ghost Observe", list("Ghost", "Stay in Body"))
+			if(response == "Ghost")
+				ghostize(FALSE)
+			else
+				return FALSE
+		else
+			ghostize(TRUE)
+			return TRUE
```
Comment (tralezab) [inline code/modules/mob/dead/observer/observer.dm:329]: check if they aren't an ethereal, and you can return the code early before doing all that tgui alert stuff. reads much nicer than making else statement pyramids

**Review by tralezab — CHANGES_REQUESTED** [review]
a tale of ifs and elses

File: code/modules/mob/dead/observer/observer.dm [hunk 3]
```diff
@@ -283,8 +283,14 @@ Works together with spawning an observer, noted above.
...
-			if(HAS_TRAIT(src, TRAIT_CORPSELOCKED) && can_reenter_corpse) //If you can re-enter the corpse you can't leave when corpselocked
-				return
+			if(HAS_TRAIT(src, TRAIT_CORPSELOCKED))
+				if(can_reenter_corpse) //If you can re-enter the corpse you can't leave when corpselocked
+					return
+				else //otherwise, continues the ghosting process
+					if(isethereal(usr)) //and cancels crystallization process for ethereals
+						var/mob/living/carbon/human/ethereal = usr
+						var/obj/item/organ/heart/ethereal/ethereal_heart = ethereal.getorganslot(ORGAN_SLOT_HEART)
+						ethereal_heart.stop_crystalization_process(ethereal)
```
Comment (ATH1909) [inline code/modules/mob/dead/observer/observer.dm:293]: non-ethereals can be revived by ethereal hearts, IIRC, and ethereals can get their hearts replaced with mechanical or human ones. the isethereal() check here should instead be replaced by a check to see if the humanoid's heart is an ethereal heart.

**Pull request merged**